sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -328,6 +333,9 @@ case class DataSource(
                 "It must be specified manually")
         }
 
+        SchemaUtils.checkSchemaColumnNameDuplication(
+          dataSchema, "the datasource", sparkSession.sessionState.conf.caseSensitiveAnalysis)
```

Comment:
I though firstly I added this check for `userSpecifiedSchema`: https://github.com/apache/spark/pull/17758/files#diff-7a6cb188d2ae31eb3347b5629a679cecR325. But, does we need to get a schema from `userSpecifiedSchema` here? As you suggested, in this code path, ISTM it's okay to get a schema only from the metadata log. I checked existing test suites though, I couldn't find tests for using `userSpecifiedSchema` in this code path.